### PR TITLE
[Calendar/Email] Add FindContact New Flow

### PR DESCRIPTION
**Pull request by zidaneymar**
Add FindContact new flow for email/calendar.
Repalce ConfirmRecipient Dialog in Calendar skill and several steps in CreateEvent Email skill.

## Testing Steps

Email: send an email to Tianyu, forward an email to Tianyu
Calendar: create a meeting with Tianyu

## Checklist
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the appropriate unit tests
- [x] I have updated related documentation

### CODE REVIEW

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/EmailSkillDialog.cs [hunk 1]
```diff
@@ -639,29 +648,197 @@ protected async Task<string> GetNameListStringAsync(WaterfallStepContext sc)
...
+            {
+                return unionList.First();
+            }
+
+            var nameString = string.Join(", ", unionList.ToArray().SkipLast(1)) + string.Format(CommonStrings.SeparatorFormat, CommonStrings.And) + unionList.Last();
+            return nameString;
+        }
+
+        protected async Task<PromptOptions> GenerateOptionsForEmail(Person confirmedPerson, ITurnContext context, bool isSinglePage = true)
+        {
```
Comment (DingmaomaoBJTU) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/EmailSkillDialog.cs]: Can this move to findcontactdialog?

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/EmailSkillDialog.cs [hunk 2]
```diff
@@ -639,29 +648,197 @@ protected async Task<string> GetNameListStringAsync(WaterfallStepContext sc)
...
+                    result += i + 1 + ": ";
+                }
+
+                result += choice.Value.Split(":").LastOrDefault() + "\r\n";
+            }
+
+            return result;
+        }
+
+        protected async Task<PromptOptions> GenerateOptionsForName(List<Person> unionList, ITurnContext context, bool isSinglePage = true)
```
Comment (DingmaomaoBJTU) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/EmailSkillDialog.cs]: >protected async Task<PromptOptions> GenerateOptionsForName(List<Person> unionList, ITurnContext context, bool isSinglePage = true) [](start = 8, length = 130)

Can this also move to findcontactdialog?

File: solutions/Virtual-Assistant/src/csharp/skills/calendarskill/calendarskill/Dialogs/FindContact/FindContactDialog.cs [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+                var skillOptions = sc.Options;
+
+                if ((state.AttendeesNameList == null) || (state.AttendeesNameList.Count == 0))
+                {
+                    return await sc.BeginDialogAsync(Actions.UpdateName);
+                }
+
+                var unionList = new List<CustomizedPerson>();
+                var emailList = new List<string>();
+                if (skillOptions != null)
```
Comment (DingmaomaoBJTU) [inline solutions/Virtual-Assistant/src/csharp/skills/calendarskill/calendarskill/Dialogs/FindContact/FindContactDialog.cs]: >if (skillOptions != null) [](start = 16, length = 25)

Should always pass options.

File: solutions/Virtual-Assistant/src/csharp/skills/calendarskill/calendarskill/Dialogs/FindContact/FindContactDialog.cs [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+                var state = await Accessor.GetAsync(sc.Context);
+                var currentRecipientName = state.AttendeesNameList[state.ConfirmAttendeesNameIndex];
+                if (state.FirstRetryInFindContact)
+                {
+                    state.FirstRetryInFindContact = false;
+                    return await sc.PromptAsync(Actions.Prompt, new PromptOptions { Prompt = sc.Context.Activity.CreateReply(FindContactResponses.UserNotFound) });
+                }
+                else
+                {
+                    return await sc.CancelAllDialogsAsync();
```
Comment (DingmaomaoBJTU) [inline solutions/Virtual-Assistant/src/csharp/skills/calendarskill/calendarskill/Dialogs/FindContact/FindContactDialog.cs:76]: >return await sc.CancelAllDialogsAsync(); [](start = 20, length = 40)

Try throw exception here?

File: solutions/Virtual-Assistant/src/csharp/skills/calendarskill/calendarskill/Dialogs/FindContact/FindContactDialog.cs [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+                        if (generalTopIntent == General.Intent.Next)
+                        {
+                            state.ShowAttendeesIndex++;
+                        }
+                        else if (generalTopIntent == General.Intent.Previous)
+                        {
+                            if (state.ShowAttendeesIndex > 0)
+                            {
+                                state.ShowAttendeesIndex--;
+                            }
```
Comment (DingmaomaoBJTU) [inline solutions/Virtual-Assistant/src/csharp/skills/calendarskill/calendarskill/Dialogs/FindContact/FindContactDialog.cs:297]: Add description for previous page of the first page, next page of the last page.

File: solutions/Virtual-Assistant/src/csharp/skills/calendarskill/calendarskill/Dialogs/FindContact/FindContactDialog.cs (hunk 4 above)
Comment (zidaneymar) [reply solutions/Virtual-Assistant/src/csharp/skills/calendarskill/calendarskill/Dialogs/FindContact/FindContactDialog.cs:76]: > > return await sc.CancelAllDialogsAsync(); [](start = 20, length = 40)
> 
> Try throw exception here?

Exception handling removed

**Review by lzc850612 — APPROVED** [review]

**Review by DingmaomaoBJTU — APPROVED** [review]

**Pull request merged**